### PR TITLE
feat(linter/eslint): Implement max-classes-per-file

### PR DESCRIPTION
**Pull request by jelly**
Rule Detail:
[link](https://eslint.org/docs/latest/rules/max-classes-per-file)

### CODE REVIEW

File: crates/oxc_linter/src/rules/eslint/max_classes_per_file.rs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+            if class_count == self.max {
+                if let AstKind::Class(class) = kind {
+                    span = class.span;
+                }
+            }
+
+            if !kind.is_declaration() && self.ignore_expressions {
+            } else {
+                class_count += 1;
+            }
```
Comment (jelly) [inline crates/oxc_linter/src/rules/eslint/max_classes_per_file.rs:101]: Maybe there is a more Rust-like way to write this?

Comment (jelly) [reply crates/oxc_linter/src/rules/eslint/max_classes_per_file.rs:101]: I could change this into `filter().cnt()` and then look up `span` via `max` but I am not 100% sure how one does it as `ClassId` is not an integer but a Struct ClassId?

Comment (Dunqing) [reply crates/oxc_linter/src/rules/eslint/max_classes_per_file.rs:101]: I think you can implement both `ignore_expression` is true or not logic separately

You can add the `len` method in classes and check `class_count <= self.max` first

```rust
let mut class_count = ctx.semantic.classes().len()
if class_count <= self.max {
   return;
}
```

If true
```rust
let ast_node_id = ctx.semantic().classes().get_node_id(self.max as ClassId)
let AstKind::Class(class) = ctx.nodes().kind(*node_id);
span = class.span;
``` 

If false, you can use `ctx.semantic().classes().iter_enumerated().filter()` to get `class_count` and first `span`. if `class_count <= self.max` we can return

**Review by anonrig — APPROVED** [review]

File: crates/oxc_linter/src/rules/eslint/max_classes_per_file.rs (hunk 1 above)
Comment (jelly) [reply crates/oxc_linter/src/rules/eslint/max_classes_per_file.rs]: Thanks for the suggestions, I implemented something along the same lines.

File: crates/oxc_linter/src/rules/eslint/max_classes_per_file.rs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+        let ast_node_id = ctx.semantic().classes().get_node_id(ClassId::from(self.max));
+        let span = if let AstKind::Class(class) = ctx.nodes().kind(ast_node_id) {
+            class.span
+        } else {
+            Span::new(0, 0)
+        };
+
+        let error =
+            format!("File has too many classes ({}). Maximum allowed is {}", class_count, self.max);
```
Comment (Dunqing) [inline crates/oxc_linter/src/rules/eslint/max_classes_per_file.rs:111]: Can you put this error message into the `max_classes_per_file_diagnostic`? That way we only have to call `format!` once

**Review by Dunqing — CHANGES_REQUESTED** [review]

**Review by Dunqing — APPROVED** [review]

**Comment by Dunqing** [comment]
Thank you!

**Pull request merged**